autonat.go

```diff
@@ -0,0 +1,170 @@
+package autonat
+
+import (
+	"context"
+	"errors"
+	"math/rand"
+	"sync"
+	"time"
+
+	host "github.com/libp2p/go-libp2p-host"
+	peer "github.com/libp2p/go-libp2p-peer"
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+type NATStatus int
+
+const (
+	// NAT status is unknown
+	NATStatusUnknown NATStatus = iota
+	// NAT status is publicly dialable
+	NATStatusPublic
+	// NAT status is private network
+	NATStatusPrivate
+)
+
+var (
+	AutoNATBootDelay       = 15 * time.Second
+	AutoNATRefreshInterval = 15 * time.Minute
+
+	AutoNATRequestTimeout = 60 * time.Second
+)
+
+// AutoNAT is the interface for ambient NAT autodiscovery
+type AutoNAT interface {
+	// Status returns the current NAT status
+	Status() NATStatus
+	// PublicAddr returns the public dial address when NAT status is public and an
+	// error otherwise
+	PublicAddr() (ma.Multiaddr, error)
+}
+
+// AmbientAutoNAT is the implementation of ambient NAT autodiscovery
+type AmbientAutoNAT struct {
```

Comment:
@vyzo noob question, is `ambient` being used in the dictionary definition sense or is there a networking specific meaning that I don't know about?

Reply:
it means that it runs in the background, with the need for the user to do anything (other than instantiating it)

Reply:
Cheers